### ISSUE TITLE
Pause when holding an arrow key

### ISSUE THREAD

**Issue by j-hudecek** [openseadragon/openseadragon#1609]
When pressing an arrow the animation to move starts immediately - that is great. But when you want to move further and you hold the arrow button there is a significant delay before it starts.

I understand it's a system behavior on how the keypresses are processed. Workaround would be to have our own little game-like loop which triggers the event as long as keyup doesn't happen, like here:
https://stackoverflow.com/questions/11197220/javascript-move-delay-and-multiple-keystrokes

Is this something you've considered?

**Comment by iangilman** [comment]
Yes, great idea! I think this is the first time it's been mentioned, but I agree it would be a nice improvement. 

Let me know if you're interested in working on this, and I can help point you to the right part of the code.

**Comment by MichaelYogar** [comment]
Hi, I can try to work on this and submit a pull request!

**Comment by MichaelYogar** [comment]
@iangilman if you can point me to the right part of the code as you suggested, I can try to fix this. Thanks!

**Comment by iangilman** [comment]
@MichaelYogar Great! Apologies for the delayed response… If you're still interested, here's the info…

You'll want to give the viewer an object property to keep track of which keys are pressed. That can be done somewhere in the viewer's constructor:

https://github.com/openseadragon/openseadragon/blob/4ab8cc748188f6c4dc5ad7d23e91945dd0194173/src/viewer.js#L65

Then you can update the "key down" handler to modify the "keys pressed" object:

https://github.com/openseadragon/openseadragon/blob/4ab8cc748188f6c4dc5ad7d23e91945dd0194173/src/viewer.js#L2536

You'll need to make a new "key up" handler and hook it up here:

https://github.com/openseadragon/openseadragon/blob/4ab8cc748188f6c4dc5ad7d23e91945dd0194173/src/viewer.js#L289

Then every time through the update loop you can check the keys and pan the viewer as appropriate:

https://github.com/openseadragon/openseadragon/blob/4ab8cc748188f6c4dc5ad7d23e91945dd0194173/src/viewer.js#L3280

Be sure to honor the functionality that's already built in the "key down" handler. Also note that we may need to adjust the speed once it's happening every frame, so it's not too fast.

As you can see, this is probably more work than a simple "good first issue", though I don't think it requires understanding of the rest of the code, so it is a good introduction to the code base if you're up for it. Let me know if you have additional questions!

**Comment by MichaelYogar** [comment]
@iangilman Hey, thanks for your reply! I'm still interested in doing this for sure. If this is urgent, you might want to assign it to someone else. If not, I can give this a shot around next month because I'm busy now.

**Comment by HamzaTatheer** [comment]
Could this be done in the following way, update pressed_key in OnCanvasKeyDown, set pressed_key to null onCanvasKeyUp (if the release key is the same as stored pressed key). And then move all the onCanvasKeyPress switch case statement code to updateOnce.

Regarding, speed we can add a variable that lets us control how many events per second will be detected. This way user has control over how fast the controls work. Do you totally agree with this approach or not

**Comment by iangilman** [comment]
@HamzaTatheer Yes, that seems like a good approach! The pressed key variable should be an object, with each key code being a property key in the object... That way we can keep track of multiple keys down at the same time (so you can hold both W and A to move diagonally). 

Also, it would be good to keep track of when the user leaves the window and clean out all of the pressed keys, since we don't get keyup events when our window isn't focused. I think you can get that by watching the `blur` event on `window`.

**Comment by HamzaTatheer** [comment]
Alright Makes Sense, I want to know that can two people complete this at the same time. How could they collaborate. After forking, What should be the flow

**Comment by HamzaTatheer** [comment]
okay I will try to work on this. Seems like a relatively easy task. If I don't complete all of it, Can someone carry on by contributing to the forked repo that I have

**Comment by iangilman** [comment]
@HamzaTatheer Great! And yes; if you start a PR, but can't finish it, or if you link to the branch from here, it can be a starting point for the next person.

**Comment by msalsbery** [comment]
Anyone working on this or is there a stalled PR?

**Comment by iangilman** [comment]
Not that I'm aware of. Did you make any progress on this, @HamzaTatheer?

**Comment by achu1998** [comment]
@iangilman, if no one is working on this, can I look at it?

**Comment by iangilman** [comment]
Fixed by #2735
